cmd/clusterctl/pkg/client/cluster/inventory.go

```diff
@@ -294,3 +308,115 @@ func (p *inventoryClient) GetDefaultProviderNamespace(provider string) (string,
 	// There is no provider or more than one namespace for this provider; in both cases, a default provider namespace cannot be decided.
 	return "", nil
 }
+
+func (p *inventoryClient) GetManagementGroups() ([]ManagementGroup, error) {
+	providers, err := p.List()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if there are core providers with overlapping watching namespaces, if yes, it is not possible to define the management groups, so return error
+	if err := checkOverlappingCoreProviders(providers); err != nil {
+		return nil, err
+	}
+
+	// check if a Bootstrap/ControlPlane/Infrastructure providers can't be combined with any core provider or
+	// if any Bootstrap/ControlPlane/Infrastructure provider is watching for objects controlled by more than one core provider.
+	// In both cases it is not possible to define the management groups, so return error
+	if err := checkOverlappingProviders(providers); err != nil {
+		return nil, err
+	}
+
+	managementGroups := []ManagementGroup{}
+	for _, coreProvider := range providers {
+		if coreProvider.Type != string(clusterctlv1.CoreProviderType) {
+			continue
+		}
+
+		group := ManagementGroup{CoreProvider: coreProvider}
+		for _, provider := range providers {
+			if coreProvider.WatchedNamespace == "" || coreProvider.WatchedNamespace == provider.WatchedNamespace || provider.WatchedNamespace == "" {
```

Comment:
Should this be true either if core and provider are watching the same namespace or if they're both watching all?

If so, I think we can simplify using only `coreProvider.WatchedNamespace == provider.WatchedNamespace`.

Reply:
Your proposal does not identify as overlapping A watching all namespaces, B watching a namespace (or vice versa).

Two providers have overlapping watching namespace if they are watching the same namespace or 
at least one is Watching all namespaces is watching all namespaces